### PR TITLE
CA - Last hit quadruplet chi2 filter

### PR DESCRIPTION
**Pull request by felicepantaleo**
Adding the possibility to have a filter on the last hit of a quadruplet, like in quadruplets by propagation.
If two quadruplets have the first three hits in common and the fourth hits are on the same layer, pick the one with the lower chi2. 

This has the effect of reducing the fake and duplicate rate within the same iteration: e.g. https://fpantale.web.cern.ch/fpantale/offline_tuning/lastLayerChi2FilterPR/plots_initialStepPreSplitting.html

blue Quadruplets by propagation
red CA in the release
black CA with filter turned on (this PR) 

@VinInn @rovere @makortel

### CODE REVIEW

**Comment by felicepantaleo** [comment]
@mtosi @JanFSchulte 
This may be interesting for you as well.

**Comment by VinInn** [comment]
@cmsbuild , please test

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc [hunk 1]
```diff
@@ -278,6 +300,27 @@ void CAHitQuadrupletGenerator::hitQuadruplets(const TrackingRegion& region,
...
 
+    if(caOnlyOneLastHitPerLayerFilter)
+    {
+            fourthLayerId = tTopo->layer(ahit->geographicalId());
+            sideId = tTopo->side(ahit->geographicalId());
+            subDetId = ahit->geographicalId().subdetId();
+	    isTheSameTriplet = (quadId != 0) && (foundQuadruplets[quadId][0]->getCellId() ==  previousCellIds[0]) && (foundQuadruplets[quadId][1]->getCellId() ==  previousCellIds[1]);
+            isTheSameFourthLayer = (quadId != 0) &&  (fourthLayerId == previousfourthLayerId) && (subDetId == previousSubDetId) && (sideId == previousSideId);
+
+	    previousCellIds = {{foundQuadruplets[quadId][0]->getCellId(), foundQuadruplets[quadId][1]->getCellId()}};
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: some reduction of copy-paste, especially inplace multi-operator calls would be useful

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17083 9cb181c
- code changes are in line with description; the new feature is not enabled by default
- jenkins tests pass and comparisons show no differences (CA is not in the RECO step yet, so, no effect is expected anyways)

Based on the supplied MTV plots, the effect is pretty clear at the seeding level.
For tracks the effect is marginal.

wf10224ttfromfp_seed_effandfake_pt

**Comment by mtosi** [comment]
why don't we want to have the OnlyOneLastHitPerLayerFilter set to True as default ?
@felicepantaleo @rovere @VinInn @makortel @JanFSchulte

**Comment by felicepantaleo** [comment]
@mtosi it's false if you don't specify that you want to use it in the configuration.
The "philosophical reason" for that is that it goes against the concept of the CA.

**Comment by makortel** [comment]
@mtosi Setting it optional now was my request for mostly to be cautious (even if in general I'm in favour of it).

The PR adding workflows for CA in offline (#16911) indicated a problem in b tagging that we are currently investigating. I'd prefer to not change the offline-default CA parameters until the origin of the problem is fully understood (or it is demonstrated that the change cures the problem).

I mainly feel that in addition to Felice's tests the feature would benefit from some further studies (for both offline and HLT) before switching it on. Integrating the feature as default-off allows Felice/us to integrate it now without affecting anything. OTOH, if HLT prefers to switch it on by default, that's fine for me (as long as it is only for HLT for now).

**Comment by mtosi** [comment]
I see, thanks !
  mia

**Comment by davidlange6** [comment]
+1

**Pull request merged**